### PR TITLE
fix get dataloader size

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

smhow the master is still failing

Fixes #2293 #2307 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -285,7 +285,11 @@ def _reset_eval_dataloader(
         # datasets could be none, 1 or 2+
         if len(dataloaders) != 0:
             for i, dataloader in enumerate(dataloaders):
-                num_batches = 0
+                try:
+                    num_batches = len(dataloader)
+                except (TypeError, NotImplementedError):
+                    num_batches = float('inf')
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:291]:
```suggestion
                num_batches = len(dataloader) if callable(getattr(dataloader, "__len__")) else float("inf")
```
Wouldn't that be better?

File: tests/trainer/test_lr_finder.py [hunk 2]
```diff
@@ -155,7 +154,7 @@ def test_accumulation_and_early_stopping(tmpdir):
         'Learning rate was not altered after running learning rate finder'
     assert len(lrfinder.results['lr']) == 100, \
         'Early stopping for learning rate finder did not work'
-    assert lrfinder._total_batch_idx == 100 * 2, \
+    assert lrfinder._total_batch_idx == 190, \
```
Comment (awaelchli) [inline tests/trainer/test_lr_finder.py:157]: how is this related?

File: pytorch_lightning/trainer/data_loading.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:291]: I think it is bad to use exceptions for control flow like this. If the user implements `len` but has a type error, we catch it and the program crashes somewhere else, making it much harder for the user to debug.

File: tests/trainer/test_lr_finder.py (hunk 2 above)
Comment (Borda) [reply tests/trainer/test_lr_finder.py:157]: not really but the test was skipped completely

File: pytorch_lightning/trainer/data_loading.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/data_loading.py]: seems like the suggestion did not work, so reverting it back... :(

File: tests/trainer/test_lr_finder.py (hunk 2 above)
Comment (awaelchli) [reply tests/trainer/test_lr_finder.py:157]: :(

File: pytorch_lightning/trainer/data_loading.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:291]: :(

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:291]: what was the error?

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:291]: what about `num_batches = len(dataloader) if _has_len(dataloader) else float('inf')`?

Comment (thschaaf) [reply pytorch_lightning/trainer/data_loading.py:291]: This seems to be a longterm solution.

Comment (Borda) [reply pytorch_lightning/trainer/data_loading.py:291]: @awaelchli see here - https://github.com/PyTorchLightning/pytorch-lightning/runs/812000184

Comment (Borda) [reply pytorch_lightning/trainer/data_loading.py:291]: the problem is raising an exception in the `__len__`

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:291]: yeah my orignal suggestions was wrong sorry. my new suggestion is to use the already existing _has_len method
`num_batches = len(dataloader) if _has_len(dataloader) else float('inf')`
but not a big deal, would just be cleaner.
